File: mobile/lib/features/cos_user_context/cos_user_context.dart
import 'dart:convert';

import 'package:nostr/nostr.dart' as nostr;

import '../../shared/relay/nostr_models.dart';

const cosUserContextSchema = 'mac-workspace/cos-user-context/v1';
const cosWorkspaceModules = {
  'today',
  'my_actions',
  'messages',
  'assistant',
  'running_order',
  'agents',
};

class CosUserContext {
  final String eventId;
  final String channelId;
  final String assigneePubkey;
  final List<String> modules;
  final int createdAt;

  const CosUserContext({
    required this.eventId,
    required this.channelId,
    required this.assigneePubkey,
    required this.modules,
    required this.createdAt,
  });

  bool hasModule(String module) => modules.contains(module);
  bool get canUseControlRoom =>
      hasModule('agents') && hasModule('running_order');
}

CosUserContext parseCosUserContext(
  NostrEvent event, {
  required String expectedAssignee,
  String? expectedChannelId,
}) {
  if (!_hasValidSignature(event)) {
    throw const FormatException('COS user-context signature is invalid');
  }
  if (event.kind != EventKind.cosUserContext) {
    throw const FormatException('Expected a COS user-context event');
  }
  final assignee = _exactTag(event, 'p').toLowerCase();
  if (assignee != expectedAssignee.toLowerCase()) {
    throw const FormatException('COS user context belongs to another identity');
  }
  if (_exactTag(event, 'd') != 'context:$assignee') {
    throw const FormatException('COS user-context coordinate is invalid');
  }
  final channelId = _exactTag(event, 'h');
  if (expectedChannelId != null && channelId != expectedChannelId) {
    throw const FormatException('COS user context belongs to another channel');
  }
  final decoded = jsonDecode(event.content);
  if (decoded is! Map || decoded['schema'] != cosUserContextSchema) {
    throw const FormatException('Unsupported COS user-context schema');
  }
  final rawModules = decoded['modules'];
  if (rawModules is! List ||
      rawModules.any(
        (module) => module is! String || !cosWorkspaceModules.contains(module),
      )) {
    throw const FormatException('Unsupported Workspace module');
  }
  final modules = rawModules.cast<String>();
  if (modules.toSet().length != modules.length ||
      !modules.contains('today') ||
      !modules.contains('my_actions') ||
      !modules.contains('messages')) {
    throw const FormatException('COS user-context modules are invalid');
  }
  return CosUserContext(
    eventId: event.id,
    channelId: channelId,
    assigneePubkey: assignee,
    modules: List.unmodifiable(modules),
    createdAt: event.createdAt,
  );
}

CosUserContext? selectLatestCosUserContext(
  Iterable<NostrEvent> events, {
  required String assigneePubkey,
  required String trustedBridgePubkey,
  String? expectedChannelId,
}) {
  CosUserContext? latest;
  for (final event in events) {
    if (event.pubkey.toLowerCase() != trustedBridgePubkey.toLowerCase()) {
      continue;
    }
    try {
      final candidate = parseCosUserContext(
        event,
        expectedAssignee: assigneePubkey,
        expectedChannelId: expectedChannelId,
      );
      if (latest == null ||
          candidate.createdAt > latest.createdAt ||
          (candidate.createdAt == latest.createdAt &&
              candidate.eventId.compareTo(latest.eventId) < 0)) {
        latest = candidate;
      }
    } on FormatException {
      // Ignore malformed, incorrectly scoped or untrusted projections.
    }
  }
  return latest;
}

List<String> cosUserContextChannelCandidates(
  Iterable<NostrEvent> events, {
  required String assigneePubkey,
  required String trustedBridgePubkey,
}) {
  final channels = <String>{};
  for (final event in events) {
    if (event.pubkey.toLowerCase() != trustedBridgePubkey.toLowerCase()) {
      continue;
    }
    try {
      channels.add(
        parseCosUserContext(event, expectedAssignee: assigneePubkey).channelId,
      );
    } on FormatException {
      // Only structurally valid bridge projections may nominate a channel.
    }
  }
  return channels.toList()..sort();
}

/// Returns the one current private COS identity channel in this relay
/// workspace. Ambiguous, stale or malformed membership fails closed.
String? resolveAuthoritativeCosUserContextChannel({
  required Iterable<String> candidateChannelIds,
  required Iterable<NostrEvent> metadataEvents,
  required Iterable<NostrEvent> membershipEvents,
  required String assigneePubkey,
  required String trustedBridgePubkey,
}) {
  final metadata = _latestByCoordinate(metadataEvents, 39000);
  final memberships = _latestByCoordinate(membershipEvents, 39002);
  final valid = candidateChannelIds.toSet().where((channelId) {
    final channelMetadata = metadata[channelId];
    final channelMembership = memberships[channelId];
    return channelMetadata != null &&
        _isPrivateMetadata(channelMetadata) &&
        channelMembership != null &&
        _hasExactFollowUpMembership(
          channelMembership,
          assigneePubkey,
          trustedBridgePubkey,
        );
  }).toList();
  return valid.length == 1 ? valid.single : null;
}

Map<String, NostrEvent> _latestByCoordinate(
  Iterable<NostrEvent> events,
  int expectedKind,
) {
  final latest = <String, NostrEvent>{};
  for (final event in events) {
    if (event.kind != expectedKind || !_hasValidSignature(event)) continue;
    String coordinate;
    try {
      coordinate = _exactTag(event, 'd');
    } on FormatException {
      continue;
    }
    final current = latest[coordinate];
    if (current == null ||
        event.createdAt > current.createdAt ||
        (event.createdAt == current.createdAt &&
            event.id.compareTo(current.id) < 0)) {
      latest[coordinate] = event;
    }
  }
  return latest;
}

bool _isPrivateMetadata(NostrEvent event) => event.tags.any(
  (tag) =>
      tag.isNotEmpty &&
      (tag[0] == 'private' ||
          (tag.length > 1 && tag[0] == 'visibility' && tag[1] == 'private')),
);

bool _hasExactFollowUpMembership(
  NostrEvent event,
  String assigneePubkey,
  String trustedBridgePubkey,
) {
  final roles = <String, String>{};
  for (final tag in event.tags) {
    if (tag.length < 2 || tag[0] != 'p') continue;
    final pubkey = tag[1].toLowerCase();
    if (roles.containsKey(pubkey)) return false;
    roles[pubkey] = tag.length > 3 && tag[3].isNotEmpty ? tag[3] : 'member';
  }
  final assignee = assigneePubkey.toLowerCase();
  final bridge = trustedBridgePubkey.toLowerCase();
  return assignee != bridge &&
      roles.length == 2 &&
      roles[bridge] == 'owner' &&
      roles[assignee] == 'member';
}

bool _hasValidSignature(NostrEvent event) {
  try {
    nostr.Event(
      event.id,
      event.pubkey,
      event.createdAt,
      event.kind,
      event.tags.map((tag) => [...tag]).toList(),
      event.content,
      event.sig,
      verify: true,
    );
    return true;
  } on Object {
    return false;
  }
}

String _exactTag(NostrEvent event, String name) {
  final values = [
    for (final tag in event.tags)
      if (tag.length > 1 && tag[0] == name) tag[1],
  ];
  if (values.length != 1 || values.single.isEmpty) {
    throw FormatException(
      'COS user context must contain exactly one $name tag',
    );
  }
  return values.single;
}

File: desktop/src/features/cos-user-context/lib/cosUserContext.ts
import type { RelayEvent } from "@/shared/api/types";
import { KIND_COS_USER_CONTEXT } from "@/shared/constants/kinds";
import { verifyEvent } from "nostr-tools/pure";

export const COS_USER_CONTEXT_SCHEMA = "mac-workspace/cos-user-context/v1";

export type CosWorkspaceModule =
  | "today"
  | "my_actions"
  | "messages"
  | "assistant"
  | "running_order"
  | "agents";

export type CosUserContext = {
  eventId: string;
  authorPubkey: string;
  channelId: string;
  assigneePubkey: string;
  tenantSlug: string;
  user: {
    id: string | number | boolean;
    name: string;
    role: string;
    roleLabel: string;
  };
  modules: CosWorkspaceModule[];
  assistant: {
    key: "mac-assistant";
    label: string;
    execution: "brain-vps";
    memoryScope: "private-channel";
  } | null;
  generatedAt: string;
  createdAt: number;
};

type JsonRecord = Record<string, unknown>;

const VALID_MODULES = new Set<CosWorkspaceModule>([
  "today",
  "my_actions",
  "messages",
  "assistant",
  "running_order",
  "agents",
]);
const REQUIRED_MODULES = ["today", "my_actions", "messages"] as const;
const TENANT_SLUG = /^[a-z0-9]+(?:-[a-z0-9]+)*$/;

function record(value: unknown): JsonRecord {
  if (value === null || typeof value !== "object" || Array.isArray(value)) {
    throw new Error("Expected an object");
  }
  return value as JsonRecord;
}

function string(value: unknown, label: string): string {
  if (typeof value !== "string" || value.trim().length === 0) {
    throw new Error(`${label} must be a non-empty string`);
  }
  return value;
}

function exactlyOneTag(event: RelayEvent, name: string): string {
  const values = event.tags
    .filter((tag) => tag[0] === name && typeof tag[1] === "string")
    .map((tag) => tag[1]);
  if (values.length !== 1) {
    throw new Error(`COS user context must contain exactly one ${name} tag`);
  }
  return values[0];
}

function parseModules(value: unknown): CosWorkspaceModule[] {
  if (!Array.isArray(value)) {
    throw new Error("modules must be an array");
  }
  const modules = value.map((module) => {
    if (
      typeof module !== "string" ||
      !VALID_MODULES.has(module as CosWorkspaceModule)
    ) {
      throw new Error("Unsupported Workspace module");
    }
    return module as CosWorkspaceModule;
  });
  if (new Set(modules).size !== modules.length) {
    throw new Error("modules contains duplicates");
  }
  for (const required of REQUIRED_MODULES) {
    if (!modules.includes(required)) {
      throw new Error(`modules must include ${required}`);
    }
  }
  return modules;
}

export function parseCosUserContext(
  event: RelayEvent,
  expectedAssignee?: string,
): CosUserContext {
  try {
    // Verify a fresh structural copy so nostr-tools cannot reuse a cached
    // verifiedSymbol from a previously verified, subsequently mutated object.
    if (
      !verifyEvent({
        id: event.id,
        pubkey: event.pubkey,
        created_at: event.created_at,
        kind: event.kind,
        tags: event.tags.map((tag) => [...tag]),
        content: event.content,
        sig: event.sig,
      })
    ) {
      throw new Error("invalid signature");
    }
  } catch {
    throw new Error("COS user context signature is invalid");
  }
  if (event.kind !== KIND_COS_USER_CONTEXT) {
    throw new Error("Expected a COS user context event");
  }
  const channelId = exactlyOneTag(event, "h");
  const assigneePubkey = exactlyOneTag(event, "p").toLowerCase();
  if (exactlyOneTag(event, "d") !== `context:${assigneePubkey}`) {
    throw new Error("COS user context coordinate must bind the identity");
  }
  if (expectedAssignee && assigneePubkey !== expectedAssignee.toLowerCase()) {
    throw new Error("COS user context belongs to a different identity");
  }

  const raw = record(JSON.parse(event.content) as unknown);
  if (raw.schema !== COS_USER_CONTEXT_SCHEMA) {
    throw new Error("Unsupported COS user context schema");
  }
  const tenantSlug = string(raw.tenant_slug, "tenant_slug");
  if (tenantSlug.length > 63 || !TENANT_SLUG.test(tenantSlug)) {
    throw new Error("tenant_slug must be a canonical lower-case slug");
  }
  const user = record(raw.user);
  if (
    user.id === null ||
    !["string", "number", "boolean"].includes(typeof user.id)
  ) {
    throw new Error("user.id must be a JSON scalar");
  }
  const modules = parseModules(raw.modules);
  const hasAssistant = modules.includes("assistant");
  let assistant: CosUserContext["assistant"] = null;
  if (raw.assistant !== null && raw.assistant !== undefined) {
    const value = record(raw.assistant);
    if (
      value.key !== "mac-assistant" ||
      value.execution !== "brain-vps" ||
      value.memory_scope !== "private-channel"
    ) {
      throw new Error(
        "Assistant context violates the staff isolation boundary",
      );
    }
    assistant = {
      key: "mac-assistant",
      label: string(value.label, "assistant.label"),
      execution: "brain-vps",
      memoryScope: "private-channel",
    };
  }
  if (hasAssistant !== Boolean(assistant)) {
    throw new Error(
      "assistant must be present exactly when the module is enabled",
    );
  }

  return {
    eventId: event.id,
    authorPubkey: event.pubkey.toLowerCase(),
    channelId,
    assigneePubkey,
    tenantSlug,
    user: {
      id: user.id as string | number | boolean,
      name: string(user.name, "user.name"),
      role: string(user.role, "user.role"),
      roleLabel: string(user.role_label, "user.role_label"),
    },
    modules,
    assistant,
    generatedAt: string(raw.generated_at, "generated_at"),
    createdAt: event.created_at,
  };
}

export function selectLatestCosUserContext(
  events: RelayEvent[],
  assigneePubkey: string,
  trustedBridgePubkey: string,
): CosUserContext | null {
  const trustedAuthor = trustedBridgePubkey.toLowerCase();
  let latest: CosUserContext | null = null;
  for (const event of events) {
    if (event.pubkey.toLowerCase() !== trustedAuthor) continue;
    try {
      const candidate = parseCosUserContext(event, assigneePubkey);
      if (
        !latest ||
        candidate.createdAt > latest.createdAt ||
        (candidate.createdAt === latest.createdAt &&
          candidate.eventId.localeCompare(latest.eventId) < 0)
      ) {
        latest = candidate;
      }
    } catch {
      // Ignore malformed or incorrectly scoped projections.
    }
  }
  return latest;
}

export function hasCosWorkspaceModule(
  context: CosUserContext | null | undefined,
  module: CosWorkspaceModule,
): boolean {
  return context?.modules.includes(module) ?? false;
}

export function currentCosUserContext({
  data,
  isError,
  isPending,
}: {
  data: CosUserContext | null | undefined;
  isError: boolean;
  isPending: boolean;
}): CosUserContext | null {
  if (isError || isPending) return null;
  return data ?? null;
}
